In [ ]:
# Программирование на Python, БИ

## НИУ ВШЭ, 2025-26 учебный год

## Групповой проект №3. Statistics in Data Analysis


## Общая информация

__Дата выдачи:__ 30.11.2025

__Защита проекта:__ 17.12.2025

## Формат выполнения

Групповой проект представляет собой творческое инициативное групповое задание, которое выполняется студентами в командах от 3 до 6 человек. В то же время, следует помнить, что чем больше в команде людей, тем выше предъявляются требования в целом к команде, а также к выполняему ею проекту. Команды разрешается формировать из студентов разных групп.

При оценивании результатов выполнения проекта командой, в обязательном порядке проверяется и учитывается в оценке вклад ***каждого участника*** в общую работу команды. Вследствие этого, крайне важно ответственно подойти к выбору сокомандников, распределению обязанностей внутри команды, а также пресечению наличия "мертвых душ" в команде.

При необходимости, команда большинством голосов имеет право исключить человека из команды *до защиты проекта*, сообщив об этом преподавателю, в случае если такой человек никак не проявлялся в ходе выполнения проекта и являлся как раз таки той самой "мертвой душой". Помните про эту возможность, поскольку наличие "мертвых душ" может негативно повлиять на качество и оценку для всего проекта и, как следствие, в целом для всей команды!

Запись составов команд осуществляется по ссылке в таблице:

https://docs.google.com/spreadsheets/d/1PQGLfAno-ixcVf1CjrHaNXlyjllm3UXMvM1sOtRdyZo/edit?usp=sharing

## Оценивание и штрафы



Групповой проект оценивается исходя из устной защиты (презентации) командой выполненного ими проекта перед комиссией. Комиссия составляется из преподавателей курса, ассистентов курса, а также приглашенных экспертов из индустрии.

В рамках каждой защиты комиссия выставляет два типа оценок.

Первая оценка — это оценка для всего проекта в целом. Она ставится комиссией, исходя из глубины проработки командой предметной области, сложности применяемых в проекте инструментов, а также прочих аспектов, связанных с содержанием выполняемого в проекте ТЗ. При выставлении этой оценки комиссия ориентируется на критерии, более подробно представленные ниже в разделе "Критерии оценки проекта".

Вторая оценка — это индивидуальная оценка каждого участника команды. Именно эта оценка выставляется впоследствии в ведомость. Индивидуальная оценка участника команды определяется комиссией, исходя из того, насколько участник ориентируется в выполненном командой проекте, как он отвечает на адресованные персонально ему вопросы комиссии, насколько в принципе прослеживается его роль в проекте, а также исходя из других аспектов, связанных с участием конкретного человека в деятельности команды. Индивидуальная оценка не может превосходить оценку команды за проект в целом (первую оценку), однако, в случае когда участник спокойно и корректно отвечает на все адресованные ему вопросы, свободно ориентируется в материале и содержании выполненного проекта, а также когда его роль в работе команды не вызывает сомнений, — в таком случае, индивидуальная оценка участника становится равной оценке для всего проекта (т.е. в идеальном случае — вторая оценка равна первой оценке). Кроме того, в исключительных случаях, при демонстрации участником на защите превосходящих ожидания уровней понимания, участия и вовлечения в рамках проекта, комиссия может своим решением повысить индивидуальную оценку для этого участника на 0,5 или 1 балл относительно общей оценки за проект (первой оценки).

Во избежание субъективности и/или предвзятости оценки тех или иных членов комиссии в отношении тех или иных команд, а также каких-либо иных проявлений неравных условий — распределение команд между комиссиями осуществляется абсолютно рандомно с использованием средств библиотеки Numpy :)) — в качестве рэндом сида выступает дата, в которую производится распределение, например, `np.random.seed(30112025)`.

Обратите, пожалуйста, внимание на то, что бОльшая часть комиссий осуществляет приём проектов в очной форме, однако могут присутствовать и 1-2 комиссии, которые будут работать дистанционно с использованием средств видеосвязи. В случае попадания команды путем случайного распределения в комиссию, работающую в очной форме, *не допускается* какое-либо изменение формата проведения для команды на онлайн-режим, равно как и любые другие перемещения команд между комиссиями (смены комиссий), для сохранения честности и непредвзятости случайного распределения.

Явка студента на защиту проекта является обязательной; вклад студента без его присутствия не оценивается и автоматически является равным нулю. Для дистанционных комиссий, дополнительным требованием является обязательное наличие включенной камеры в течение всего времени нахождения в конференции

## Описание задания

Практика A/B-тестирования, формирования и проверки статистических гипотез лежит в основе принятия решений в любой современной компании, работающей с данными. Умение корректно провести эксперимент, проинтерпретировать результаты и сформулировать выводы — имеет непосредственное влияние на возможность роста продукта, сокращение издержек и уменьшение рисков.


Данный групповой проект состоит из **двух частей**:

1. **A/B-тестирование**: анализ результатов эксперимента, которые зафиксированы в датасете `data_ab.csv`, — проверка наличия статистически и практически значимых различий между наблюдениями в контрольной и экспериментальной группах.
2. **Проверка гипотез на выбранном датасете**: выбор одного или нескольких датасетов с сайта [UCI Machine Learning Repository](https://archive.ics.uci.edu/) либо же получение соответствующих датасетов путем парсинга/работы с API; осуществление этапа первичного анализа данных (EDA); определение не менее трех содержательных гипотез *на основе проделанного EDA*; их проверка с помощью различнных статистических тестов.


### Постановка задачи

#### Часть 1. A/B-тест

Для выполнения первой части группового проекта вам предоставляется датасет с результатами проведения некоторого A/B-тестирования в компании из сферы образования.


Датасет состоит из пяти столбцов:

* `user_id` — уникальный идентификатор пользователя;

* `timestamp` — время посещения пользователем страницы;

* `group` — группа эксперимента, к которой был случайно отнесён пользователь:

  *   `control` — контрольная группа,
  *   `treatment` — экспериментальная группа;

* `landing_page` — версия лендинга, которую увидел пользователь:

  *   `old_page` — старая версия страницы,
  *   `new_page` — новая версия страницы;

* `converted` — бинарный признак, показывающий, совершил ли пользователь целевое действие (целевым действием является регистрация на курс).








## Гипотеза 
* `converted` — БИНАРНЫЙ признак, показывающий, совершил ли пользователь целевое действие, можно сказать бинарная конверсия (зарегался/нет) 

H₀ (нулевая гипотеза) - Конверсия на переделанной странице = Конверсия на старой странице  
H₁ (альтернативная) - Конверсия на новой странице выше чем на старой 


In [34]:
from scipy import stats as st
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 
import scipy.stats as stats
from scipy.stats import t, norm

In [4]:
df = pd.read_csv('data_ab.csv')
display(df)


user_id                   timestamp      group landing_page  converted
0        851104  2025-01-21 22:11:48.556739    control     old_page          0
1        804228  2025-01-12 08:01:45.159739    control     old_page          0
2        661590  2025-01-11 16:55:06.154213  treatment     new_page          0
3        853541  2025-01-08 18:28:03.143765  treatment     new_page          0
4        864975  2025-01-21 01:52:26.210827    control     old_page          1
...         ...                         ...        ...          ...        ...
294473   751197  2025-01-03 22:28:38.630509    control     old_page          0
294474   945152  2025-01-12 00:51:57.078372    control     old_page          0
294475   734608  2025-01-22 11:45:03.439544    control     old_page          0
294476   697314  2025-01-15 01:20:28.957438    control     old_page          0
294477   715931  2025-01-16 12:40:24.467417  treatment     new_page          0

[294478 rows x 5 columns]

In [ ]:
#проверка перекрестных строк 
# crosstab считает сколько строк в датасете имеют каждую комбинацию (группа - версия страницы )
pd.crosstab(df['group'], df['landing_page'])

landing_page  new_page  old_page
group                           
control           1928    145274
treatment       145311      1965

##### Что видим? у нас некоторые получили не то что должны были увидеть, то есть control group не должна была получить new_page, а treatment group не должна была получить old_page.
##### Так как из задания дизайн эксперимента подразумевает, что группы должны быть независимы

In [25]:
# посчитаем долю некорректно распределенных пользователей 

wrong_c = 1928 / (1928 + 145274)
wrong_tr = 1965 / (1965 + 145311)

print(wrong_c)
print(wrong_tr)
# их около 1% -> очень мало -> удаляем со спокойной душой 

#точнее делаем копию старого с правильными условиями 
df = df[((df['group'] == 'control')   & (df['landing_page'] == 'old_page')) |
    ((df['group'] == 'treatment') & (df['landing_page'] == 'new_page'))].copy()

#опять проверяем  
pd.crosstab(df['group'], df['landing_page'])




0.013097648129780846
0.013342296097123767


landing_page  new_page  old_page
group                           
control              0    145274
treatment       145311         0

#### Теперь проверим дубликаты

In [28]:
# их можно спокойно посмотреть по пользователям, так как если они не уникальны значит и остальное не уникально 

dups = df.duplicated(subset=['user_id'])
df[dups]
df[df['user_id'] == 773192 ]

#попался уебок удаляем дубликат и оставляем последний, так как пользователи должны быть уник
df = df.sort_values('timestamp').drop_duplicates('user_id', keep='last')
df

user_id                   timestamp      group landing_page  converted
131228   922696  2025-01-02 13:42:05.378582  treatment     new_page          0
184884   781507  2025-01-02 13:42:15.234051    control     old_page          0
83878    737319  2025-01-02 13:42:21.786186    control     old_page          0
102717   818377  2025-01-02 13:42:26.640581  treatment     new_page          0
158789   725857  2025-01-02 13:42:27.851110  treatment     new_page          0
...         ...                         ...        ...          ...        ...
158814   799244  2025-01-24 13:41:15.325359  treatment     new_page          0
47535    808330  2025-01-24 13:41:19.152664    control     old_page          0
157787   843121  2025-01-24 13:41:44.097174  treatment     new_page          0
179072   836373  2025-01-24 13:41:52.604673    control     old_page          0
193652   920411  2025-01-24 13:41:54.460509    control     old_page          0

[290584 rows x 5 columns]

In [37]:
# разделим датасет на две группы 

control = df[df['group'] == 'control']
treatment = df[df['group'] == 'treatment']

#посчитаем метрики для последующего стат теста, а мы будем пользоваться z-тестом для пропорций , тк у нас метрика бинарная а конверсия группы = доля зареганных пользователей 

control_proportion = control['converted'].sum() / control.shape[0]
treatment_proportion = treatment['converted'].sum() / treatment.shape[0]


f_proportion = (control['converted'].sum() + treatment['converted'].sum()) / (control.shape[0] + treatment.shape[0])
 


SE = np.sqrt(f_proportion * (1 - f_proportion) * (1/control.shape[0] + 1/treatment.shape[0]))

z = (treatment_proportion - control_proportion) / SE

p_value = 2 * (1 - norm.cdf(abs(z)))

z, p_value

(np.float64(-1.3109241984234394), np.float64(0.18988337448195103))

# гпт ответ надо будет поправить : Интерпретация:

z ≈ -1.31 → конверсия treatment ниже control примерно на 1.3 стандартных ошибки (знак «минус» говорит, что новая страница чуть хуже).

p-value ≈ 0.19 > 0.05 → разница статистически незначима при уровне значимости 5%.

То есть при α = 0.05:

Мы не отвергаем нулевую гипотезу
и не можем уверенно сказать, что новая страница отличается от старой по конверсии.

По замыслу A/B-теста контрольной группе (`control`) должна была показываться старая версия лендинга — `old_page`; экспериментальной группе (`treatment`) должна была показываться новая версия лендинга — `new_page`.

Вам следует:
- понять структуру данных, провести EDA и выбрать метрику, с которой будете работать;
- проверить корректность проведения тестирования (посмотреть на однородность групп; задуматься об A/A тесте; не забыть, что в нашем распоряжении есть еще и временные метки);
- сформулировать нулевую и альтернативную гипотезы на основании выбранной метрики;
- выбрать и обосновать статистический тест;
- решить, нужно ли устаналивать значения каких-либо параметров до начала анализа тестирования;
- оценить разницу в результатах и сделать выводы о практической значимости эксперимента;
- провести дополнительный анализ при необходимости (например, с помощью [bootstrap](https://www.geeksforgeeks.org/maths/bootstrap-method/?utm_source=chatgpt.com)  или через анализ чувствительности);
- сформулировать рекомендации и бизнес-инсайты по результатам эксперимента.

**Важно**: это лишь возможные шаги для вашего исследования и воспринимать их следует как подсказки. Ваша центральная задача в любом случае представлена в общем виде — осуществить качественный и полный анализ результатов проведенного AB-тестирования.



#### Часть 2. Проверка гипотез методами математической статистики

В этой части группового проекта вам предстоит самим выбрать предметную область, а также самим получить датасет или датасеты, с которыми вы будете работать.

С точки зрения источников данных у вас есть две опции: вы можете найти ваш датасет — или 2-3 датасета в случае необходимости — на сайте [UCI Machine Learning Repository](https://archive.ics.uci.edu/) (брать датасеты необходимо именно с этого сайта); либо же вторая опция — можно собрать датасет/датасеты вручную с помощью средств Parsing'a и API. Допускается использование и комбинации этих двух опций. В любом случае, *каждый* используемый вами датасет должен (до этапа EDA) содержать **не менее 2000 строк** и **не менее 8 признаков**.




После получения датасетов вам следует:
- провести EDA с осмысленными визуализациями и качественным поиском первичных закономерностей;
- на основе проделенного EDA сформулировать **минимум три чёткие, содержательные гипотезы**;
- подобрать и реализовать **минимум три различных статистических теста** для проверки сформулированных раннее гипотез, причём каждый тест должен быть обязательно обоснован (различные статистические тесты в нашем случае означает, что статистические критерии для этих тестов должны быть разными);
- корректно и полно проинтерпретировать полученные результаты и сформулировать итоговые выводы.

**Важно:** вы должны очень ответственно подойти к выбору датасета, так как гипотезы должны быть *не случайными*, а логически вытекающими из осуществленного EDA и, кроме того, имеющими реальный интерес для изучения и проверки с точки зрения бизнес-составляющей. В случае если по одному датасету у вас не получается сформировать 3 достаточно осмысленных гипотезы — разрешается использовать суммарно вплоть до 3 датасетов. В таком случае, все эти датасеты должны иметь какую-то связь между собой.

### Общие требования к проекту

В качестве результата выполнения группового проекта ваша команда должна получить презентацию и защитить ее перед комиссией. Наличие репозитория с кодом на GitHub является необязательным условием, но приветствуется.

Оформление презентации остаётся полностью на ваше усмотрение, но помните, что результат должен быть релевантен для демонстрации бизнес-заказчику — комиссию, принимающую вашу работу, правильнее всего воспринимать именно в таком качестве. Например, вставлять в презентацию строчки кода или злоупотреблять скринами блокнота не рекомендуется.

С точки зрения концепции выполнения проекта — вам необходимо принять на себя роль дата-аналитиков: продемонстрировать способность грамотно работать с экспериментальными статистическими данными, что включает в себя проверку корректности проведения и интерпретацию результатов построенного A/B-тестирования, правильную формулировку гипотез, выбор подходящих статистических методов, моделей и их параметров, а также построение итоговых выводов из анализа — и исследователей: уметь выбирать релевантный датасет, проводить содержательный EDA, находить закономерности и превращать их в осмысленные статистические гипотезы для проверки. А в конце, конечно, — презентовать всю эту красоту в понятном виде потенциальному заказчику.

Подчеркнем отдельно: ваша задача — не просто провести тесты и сделать какие-то выводы из них, но разобраться в данных, изучить их природу, выявить реальные зависимости и инсайты, а также на основе проделанного анализа показать, почему эти находки важны для бизнеса или для выбранного контекста датасета. Разумеется, чем интереснее выбранная вами предметная область, чем осмысленнее сформулированные гипотезы и чем глубже будет проведенное вами исследование — тем выше окажется итоговое качество проекта и тем выше он будет оценён!

### Полезные замечания и советы

- Графики и визуализации помогают наглядно доносить информацию, а также анализировать данные
- Формулируйте гипотезы *после* EDA. Не придумывайте гипотезы "вслепую", сначала погрузитесь в данные, найдите закономерности, паттерны и лишь затем переходите к выделению статически проверяемых гипотез
- Не забывайте о прикладной значимости ваших проектов — ваш анализ должен быть полезен потенциальному заказчику
- Используйте разные подходы к проверке гипотез. Ищите способы усилить ваши тесты. Применяйте новые методы и идеи, а затем рассказывайте про них нам. Не бойтесь эксперементировать! Это будет только поощряться :)


## Критерии оценки проекта

Оценка за групповой проект в целом — то есть первая из двух оценок — выставляется по 10 балльной шкале, исходя из 3 критериев. Также комиссия на своё усмотрение может добавить (но не снизить) некоторое количество бонусных баллов за рамками данных критериев.

Каждый из критериев будет перечислен и рассмотрен ниже.

### Критерий 1. Анализ результатов A/B-тестирования (до 3,5 баллов)

Данный критерий оценивает качество, глубину и корректность проведенного анализа результатов A/B-тестирования — включая выбор метрики, оценку реализации теста — а также спектр статистических методов и их обоснованность, формулирование гипотез и выводов по проделанному исследованию.

*По данному критерию можно получить **1 балл и выше** только при условии осмысленного осуществления этапа EDA, корректного выбора метрики, верной формулировки гипотез, обоснованного проведения статистического теста и правильной интерпретации полученных результатов.*

*По данному критерию можно получить **2 балла и выше** только при условии учёта в значительной степени временной составляющей эксперимента.*

*По данному критерию можно получить **3 балла и выше** только при условии обоснованного применения дополнительных инструментов/подходов для анализа результатов A/B-тестирования, не упомянутых ранее в этом ноутбуке и не рассматривавшихся в рамках соответствующего семинара и лекции (bootstrap'a или анализа чувствительности может хватить на 3 балла при условии корректного и осмысленного их применения).*




### Критерий 2. Выбор предметной области и датасета, EDA, формулирование гипотез (до 3 баллов)

Данный критерий оценивает, насколько обоснованными и осмысленными получились сформулированные вами гипотезы на основании проведенного EDA по выбранной вами второй предметной области и датасетам, связанным с ней.

*__Важно:__ в случае оценки __0 баллов__ за этот критерий — оценка за следующий критерий также становится равна нулю.*

*По данному критерию нельзя получить **больше 0 баллов** при отсутствии 3 сформулированных по датасету или связанным друг с другом датасетам, гипотез, имеющих разную природу.*

*По данному критерию нельзя получить **больше 0 баллов** при условии, что хотя бы один используемый в проекте датасет не удовлетворяет требованиям, описанным в разделах выше.*

*По данному критерию можно получить **1 балл и выше** только при условии осмысленного осуществления этапа EDA и наличия 3 гипотез, напрямую логически вытекающих из этого этапа.*


### Критерий 3. Проведение статистических тестов по выбранной предметной области (до 3,5 баллов)

Данный критерий оценивает качество статистического анализа, корректность выбора тестов, разнообразие использованных статистических инструментов, интерпретацию результатов и глубину полученных выводов.

*По данному критерию можно получить **1 балл и выше** только при условии корректного использования трех различных статистических тестов.*

*По данному критерию можно получить **2 балла и выше** только при условии полностью обоснованного (и правильного) выбора каждого из используемых статистических тестов.*

*По данному критерию можно получить **3 балла и выше** только при условии проведения сравнительного анализа нескольких тестов друг с другом хотя бы для одной гипотезы, а также осмысленной интерпретации проделанного анализа в контексте реальной исследовательской задачи.*